libgnucash/engine/Account.cpp

```diff
@@ -73,8 +74,9 @@ static const std::string AB_TRANS_RETRIEVAL("trans-retrieval");
 
 static gnc_numeric GetBalanceAsOfDate (Account *acc, time64 date, gboolean ignclosing);
 
-using FinalProbabilityVec=std::vector<std::pair<std::string, int32_t>>;
-using ProbabilityVec=std::vector<std::pair<std::string, struct AccountProbability>>;
+using FinalProbabilityVec=std::vector<std::pair<int32_t, std::string>>;
```

Comment:
I was thinking about using account GUIDs (type `gnc::GUID`) here instead of account names (type `std::string`) to reduce unnecessary string allocations. Do you think, this is a good idea?

Reply:
Yes. There are a couple of other benefits: A 'gnc::GUID' is only 16 bytes while a `std::string` representation of one is 57, and `gnc::GUID::operator<` is only two comparisons while `std::string::operator<` could be up to 32 (admittedly unlikely but odds are good that it will be more than 2).

Reply:
I get a strange error, when replacing `std::string` by `gnc::GUID`:

`error: no match for ‘operator<’ (operand types are ‘const gnc::GUID’ and ‘const gnc::GUID’)`

This error occurs at lines
* 5551: `auto result = sum_of_probabilities.emplace(account.account_guid, account.probability);`
* 5755: `auto best = std::max_element(final_probabilities.begin(), final_probabilities.end());`

The first code line is in function `build_account_probabilities()` and `sum_of_probabilities` is a map, which requires to compare its keys. And the second code line is in function `gnc_account_imap_find_account_bayes()`, where the account with the maximum probability is searched, and although mainly the probabilities (first element of each pair) are compared, it can also happen, that there exist two entries with equal probability and the GUIDs (second element of each pair) have to be compared as well.

Why does the error message say, that I try to compare two `const gnc::GUID` values, although I don't use a `const` type specifier? And why isn't it possible to compare these values?

Reply:
> Why does the error message say, that I try to compare two const gnc::GUID values, although I don't use a const type specifier? And why isn't it possible to compare these values?

Because the definition of `operator<` that c++ expects takes two const refs as if it was `template <typename T> bool operator<(const T&, const T&)`. (It's not really a template, it's built into the language). C++ doesn't know how to call a member operator function in a non-class context, so we need to implement the two-operand member function.

Change the decl and impl of operator< in guid.[hc]pp to
```
bool operator<(GUID const &) const noexcept;
```
and add
```
inline bool operator<(const gnc::GUID& l, const gnc::GUID& r) { return l.operator<(r); }
```
to the bottom of guid.hpp.

Reply:
@jralls beat me to it :)

Reply:
> Change the decl and impl of operator< in guid.[hc]pp to `bool operator<(GUID const &) const noexcept;`

Done

> Add `inline bool operator<(const gnc::GUID& l, const gnc::GUID& r) { return l.operator<(r); }`

This  seems to be unnecessary. It looks like the compiler uses the `gnc::GUID` member operator function automatically. Conversely if I add this line to *guid.hpp*, I get an ambiguous overload error for `operator<`.